src/Core/Utilities/TvpHelpers.cs

```diff
@@ -0,0 +1,189 @@
+        {
+            var tvpConverter = BuildTvpConverter<T>();
+            tvpConverter.Table.SetTypeName(tableTypeName);
+            tvpConverter.AddRows(values);
+            return tvpConverter.Table;
+        }
+
+        private static TvpConverter<T> BuildTvpConverter<T>() =>
+            ((TvpConverterFactory<T>)_tvpConverterFactories[typeof(T)]).MakeConverter();
+
+        internal static TvpConverterFactory<T> BuildTvpConverterFactory<T>()
+        {
+            var table = new DataTable();
+
+            var setters = new List<Action<DataRow, T>>();
+            var ordinalGetters = new List<(int columnOrdinal, Delegate getter)>();
+            foreach (var propertyInfo in typeof(T).GetProperties(BindingFlags.Instance | BindingFlags.Public)
+                .Where(pi => Attribute.IsDefined(pi, typeof(DbOrderAttribute)))
+                .OrderBy(pi => pi.GetCustomAttribute<DbOrderAttribute>().ParameterOrder))
+            {
+                var columnType = TypeToTableType(propertyInfo.PropertyType);
+                var column = new DataColumn(propertyInfo.Name, columnType);
+
+                if (columnType == typeof(DateTime))
+                {
+                    column.DateTimeMode = DataSetDateTime.Utc;
+                }
+
+                table.Columns.Add(column);
+
+                // Create a compiled expression to get the parameter value
+                // Used in creating table rows
+                var arg = Expression.Parameter(typeof(T), "data");
+                var exprGetValue = Expression.Property(arg, propertyInfo.Name);
+                var exprConvert = Expression.Convert(exprGetValue, typeof(object));
+                var compiledGetter = Expression.Lambda<Func<T, object>>(exprConvert, arg).Compile();
+                setters.Add((row, data) => SetNullableField(row, column.Ordinal, compiledGetter(data)));
+            }
+
+            Action<T, DataTable> rowSetter = (T data, DataTable table) =>
+            {
+                var row = table.NewRow();
+                foreach (var setter in setters)
+                {
+                    setter(row, data);
+                }
+                table.Rows.Add(row);
+            };
+
+            return new TvpConverterFactory<T>(table, rowSetter);
+        }
+
+        internal static void SetNullableField(this DataRow row, int columnOrdinal, object value)
+        {
+            if (value == null)
+            {
+                row[columnOrdinal] = DBNull.Value;
+            }
+            else
+            {
+                row[columnOrdinal] = value;
+            }
+        }
+
+        private static Type TypeToTableType(Type type)
+        {
+            var underlyingType = Nullable.GetUnderlyingType(type);
+            if (underlyingType != null)
+            {
+                return underlyingType;
+            }
+            if (type.IsEnum)
+            {
+                return typeof(byte);
+            }
```

Comment:
So I know this helper class is only expecting `OrganizationUser` currently. But you do have a few enums that aren't backed by `byte`. My quick look of non byte enums are: 
| Enum                            | Type |
| ------------------------ | ----- |
| `EventType`                    | `int`  |
| `FileUploadType`            | `int`  |
| `ReferenceEventSource`  | `int`  |
| `ReferenceEventType`     | `int`  |

I don't know if any of those enums are used as properties on any objects that would ever go through here but in case they do it could be worth making a note that below could be used if any objects that goes through here would require it. 

```suggestion
            if (type.IsEnum)
            {
                return type.GetEnumUnderlyingType();
            }
```

I'm not saying that it's worth calling that yet since `typeof(byte)` should work and should save some time. I don't know how much time it would save though because I lost the trace of what that call does [here](https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs#L26). Just wanted to bring up the possibility.

Reply:
This is an interesting catch.

You're right, of course, but it has more to do with how the enum is backed in sql than in c# so really a c# method might not be good enough.

In reality, it's very likely we'll just tank all this stuff and go for a per-object extension method. Hopefully Entity Framework will save me here, but I've been saying that a lot lately so I wanted to try for a more flexible solution first.

Reply:
Yeah it's backing data type matters much more. I didn't really feel like seeing if those enums are used anywhere in the database and if so what they are backed by 🤣. So that's why I stopped short of recommending it. 

And yeah, I saw that you guys are making moves toward EF. I hope it's helpful to you guys.